File: exercises/01-bigquery-python/ex01_run_query.py
"""
Exercise 01 — Run a BigQuery Query and Print Results
-----------------------------------------------------
Goal:
    - Connect to BigQuery using the Python client library
    - Run a SQL query against a public dataset
    - Iterate over results and print rows
    - Load results into a Pandas DataFrame

Dataset used: bigquery-public-data.samples.shakespeare
"""

from google.cloud import bigquery
import pandas as pd

# ---------------------------------------------------------------------------
# STEP 1: Create a BigQuery client
# The client automatically uses GOOGLE_CLOUD_PROJECT and
# GOOGLE_APPLICATION_CREDENTIALS environment variables.
# ---------------------------------------------------------------------------
client = bigquery.Client()


# ---------------------------------------------------------------------------
# EXERCISE 1a: Run a basic query and iterate over results
# ---------------------------------------------------------------------------
def run_basic_query() -> None:
    query = """
        SELECT
            word,
            SUM(word_count) AS total_count
        FROM
            `bigquery-public-data.samples.shakespeare`
        GROUP BY
            word
        ORDER BY
            total_count DESC
        LIMIT 10
    """

    print("=== Top 10 most frequent words in Shakespeare ===")
    # TODO: Execute the query using client.query(query)
    # TODO: Call .result() to wait for completion
    # TODO: Iterate and print each row's word and total_count
    query_job = client.query(query)
    results = query_job.result()

    for row in results:
        print(f"  {row.word:<20} {row.total_count:>8,}")


# ---------------------------------------------------------------------------
# EXERCISE 1b: Load results into a Pandas DataFrame
# ---------------------------------------------------------------------------
def run_query_to_dataframe() -> pd.DataFrame:
    query = """
        SELECT
            corpus,
            COUNT(DISTINCT word) AS unique_words,
            SUM(word_count)      AS total_words
        FROM
            `bigquery-public-data.samples.shakespeare`
        GROUP BY
            corpus
        ORDER BY
            total_words DESC
        LIMIT 20
    """

    # TODO: Use client.query(query).to_dataframe() to get a DataFrame
    df = client.query(query).to_dataframe()
    print("\n=== Words per Shakespeare work (top 20) ===")
    print(df.to_string(index=False))
    return df


# ---------------------------------------------------------------------------
# EXERCISE 1c: Inspect job metadata — bytes scanned, duration
# ---------------------------------------------------------------------------
def inspect_job_metadata() -> None:
    query = """
        SELECT word, word_count
        FROM `bigquery-public-data.samples.shakespeare`
        WHERE word = 'love'
        LIMIT 100
    """

    query_job = client.query(query)
    query_job.result()  # Wait for completion

    # TODO: Print the following job statistics:
    #   - query_job.total_bytes_processed  (bytes scanned)
    #   - query_job.total_bytes_billed     (bytes billed — min 10MB)
    #   - query_job.slot_millis            (slot-milliseconds used)
    #   - (query_job.ended - query_job.started).total_seconds()
    print("\n=== Job metadata ===")
    print(f"  Bytes scanned : {query_job.total_bytes_processed:,}")
    print(f"  Bytes billed  : {query_job.total_bytes_billed:,}")
    print(f"  Slot-ms used  : {query_job.slot_millis:,}")
    duration = (query_job.ended - query_job.started).total_seconds()
    print(f"  Duration      : {duration:.2f}s")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    run_basic_query()
    run_query_to_dataframe()
    inspect_job_metadata()

# ---------------------------------------------------------------------------
# CHALLENGES (try these after the basics work)
# ---------------------------------------------------------------------------
# 1. Change the query in run_basic_query() to filter only words longer than
#    5 characters. How does the result change?
#
# 2. In inspect_job_metadata(), run the same query twice in a row.
#    Does the second run show 0 bytes billed? (BigQuery result cache)
#
# 3. Add a LIMIT 5 to the basic query. Does bytes_processed change?
#    (It should NOT — LIMIT does not reduce scan cost.)

File: exercises/02-bigquery-go/ex01_run_query.go
// Exercise 01 — Run a BigQuery Query in Go
// ------------------------------------------
// Goal:
//   - Connect to BigQuery using the Go client library
//   - Run a SQL query against a public dataset
//   - Iterate over results using an iterator
//   - Inspect job statistics (bytes processed, slot milliseconds)
//
// Run:
//   go mod init bq-go-exercises
//   go get cloud.google.com/go/bigquery
//   export GOOGLE_CLOUD_PROJECT=your-project-id
//   go run ex01_run_query.go

package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"time"

	"cloud.google.com/go/bigquery"
	"google.golang.org/api/iterator"
)

// ---------------------------------------------------------------------------
// Result struct — maps to the query output columns
// ---------------------------------------------------------------------------
type WordCount struct {
	Word       string `bigquery:"word"`
	TotalCount int64  `bigquery:"total_count"`
}

type CorpusSummary struct {
	Corpus      string `bigquery:"corpus"`
	UniqueWords int64  `bigquery:"unique_words"`
	TotalWords  int64  `bigquery:"total_words"`
}

func main() {
	ctx       := context.Background()
	projectID := os.Getenv("GOOGLE_CLOUD_PROJECT")
	if projectID == "" {
		log.Fatal("GOOGLE_CLOUD_PROJECT environment variable not set")
	}

	// ---------------------------------------------------------------------------
	// STEP 1: Create a BigQuery client
	// The client automatically picks up Application Default Credentials.
	// ---------------------------------------------------------------------------
	client, err := bigquery.NewClient(ctx, projectID)
	if err != nil {
		log.Fatalf("bigquery.NewClient: %v", err)
	}
	defer client.Close()

	// ---------------------------------------------------------------------------
	// EXERCISE 1a: Run a basic query and iterate results
	// ---------------------------------------------------------------------------
	runBasicQuery(ctx, client)

	// ---------------------------------------------------------------------------
	// EXERCISE 1b: Query with struct row binding
	// ---------------------------------------------------------------------------
	runQueryWithStructs(ctx, client)

	// ---------------------------------------------------------------------------
	// EXERCISE 1c: Inspect job metadata
	// ---------------------------------------------------------------------------
	inspectJobMetadata(ctx, client)
}

// ---------------------------------------------------------------------------
// Exercise 1a — Iterate over raw RowIterator values
// ---------------------------------------------------------------------------
func runBasicQuery(ctx context.Context, client *bigquery.Client) {
	query := client.Query(`
		SELECT
			word,
			SUM(word_count) AS total_count
		FROM
			` + "`bigquery-public-data.samples.shakespeare`" + `
		GROUP BY word
		ORDER BY total_count DESC
		LIMIT 10
	`)

	// TODO: Run the query
	// it, err := query.Read(ctx)
	it, err := query.Read(ctx)
	if err != nil {
		log.Fatalf("query.Read: %v", err)
	}

	fmt.Println("=== Top 10 most frequent words ===")
	for {
		var row WordCount
		// TODO: Call it.Next(&row) — returns iterator.Done when finished
		err := it.Next(&row)
		if err == iterator.Done {
			break
		}
		if err != nil {
			log.Fatalf("it.Next: %v", err)
		}
		fmt.Printf("  %-20s %8d\n", row.Word, row.TotalCount)
	}
}

// ---------------------------------------------------------------------------
// Exercise 1b — Use struct-based row binding for cleaner code
// ---------------------------------------------------------------------------
func runQueryWithStructs(ctx context.Context, client *bigquery.Client) {
	query := client.Query(`
		SELECT
			corpus,
			COUNT(DISTINCT word) AS unique_words,
			SUM(word_count)      AS total_words
		FROM ` + "`bigquery-public-data.samples.shakespeare`" + `
		GROUP BY corpus
		ORDER BY total_words DESC
		LIMIT 10
	`)

	it, err := query.Read(ctx)
	if err != nil {
		log.Fatalf("query.Read: %v", err)
	}

	fmt.Println("\n=== Words per Shakespeare work (top 10) ===")
	for {
		var row CorpusSummary
		err := it.Next(&row)
		if err == iterator.Done {
			break
		}
		if err != nil {
			log.Fatalf("it.Next: %v", err)
		}
		fmt.Printf("  %-40s unique=%6d  total=%8d\n",
			row.Corpus, row.UniqueWords, row.TotalWords)
	}
}

// ---------------------------------------------------------------------------
// Exercise 1c — Inspect job statistics
// ---------------------------------------------------------------------------
func inspectJobMetadata(ctx context.Context, client *bigquery.Client) {
	q := client.Query(`
		SELECT word, word_count
		FROM ` + "`bigquery-public-data.samples.shakespeare`" + `
		WHERE word = 'love'
		LIMIT 100
	`)

	// Run the query and get the Job object (not just the iterator)
	// TODO: Use q.Run(ctx) to get a *bigquery.Job
	job, err := q.Run(ctx)
	if err != nil {
		log.Fatalf("q.Run: %v", err)
	}

	// TODO: Wait for the job to complete using job.Wait(ctx)
	status, err := job.Wait(ctx)
	if err != nil {
		log.Fatalf("job.Wait: %v", err)
	}
	if err := status.Err(); err != nil {
		log.Fatalf("job failed: %v", err)
	}

	// TODO: Read job.LastStatus().Statistics to get bytes processed etc.
	stats := job.LastStatus().Statistics
	qStats := stats.Details.(*bigquery.QueryStatistics)

	fmt.Println("\n=== Job metadata ===")
	fmt.Printf("  Bytes processed : %d\n", qStats.TotalBytesProcessed)
	fmt.Printf("  Bytes billed    : %d\n", qStats.TotalBytesBilled)
	fmt.Printf("  Slot ms         : %d\n", qStats.SlotMillis)
	fmt.Printf("  Duration        : %v\n", time.Duration(stats.EndTime.Sub(stats.StartTime)))
}

// ---------------------------------------------------------------------------
// CHALLENGES
// ---------------------------------------------------------------------------
// 1. Add a goroutine-based producer: run 3 different queries concurrently
//    using goroutines and a WaitGroup. Print results as they arrive.
//
// 2. Modify runBasicQuery to accept a word length filter parameter
//    and use a parameterized query (query.Parameters field).
//
// 3. Add retries: if query.Read returns a transient error, retry up to 3 times
//    with exponential backoff using time.Sleep(2^attempt * time.Second).
